mozetl/taar/taar_similarity.py

```diff
@@ -0,0 +1,181 @@
+           addon.foreign_install:
+            return False
+
+        return True
+
+    # Create an addons dataset unnesting the addon map from each
+    # user to a list of addon GUIDS. Also filter undesired addons.
+    return (
+        users_df.rdd
+        .map(lambda p: (p["client_id"],
+             [guid for guid, data in p["active_addons"].items() if is_valid_addon(data)]))
+        .filter(lambda p: len(p[1]) > minimum_addons_count)
+        .toDF()
+        .selectExpr("_1 as client_id", "_2 as addon_ids")
+    )
+
+
+def compute_clusters(addons_df, num_clusters, random_seed):
+    """ Performs user clustering by using addon ids as features.
+    """
+
+    # Build the stages of the pipeline. We need hashing to make the next
+    # steps work.
+    hashing_stage = HashingTF(inputCol="addon_ids", outputCol="hashed_features")
+    idf_stage = IDF(inputCol="hashed_features", outputCol="features", minDocFreq=1)
+    # As a future improvement, we may add a sane value for the minimum cluster size
+    # to BisectingKMeans (e.g. minDivisibleClusterSize). For now, just make sure
+    # to pass along the random seed if needed for tests.
+    kmeans_kwargs = {"seed": random_seed} if random_seed else {}
+    bkmeans_stage = BisectingKMeans(k=num_clusters, **kmeans_kwargs)
+    pipeline = Pipeline(stages=[hashing_stage, idf_stage, bkmeans_stage])
+
+    # Run the pipeline and compute the results.
+    model = pipeline.fit(addons_df)
+    return (
+        model
+        .transform(addons_df)
+        .select(["client_id", "prediction"])
+    )
+
+
+def get_donor_pools(users_df, clusters_df, num_donors, random_seed=None):
+    """ Samples users from each cluster.
+    """
+    cluster_population = clusters_df.groupBy("prediction").count().collect()
+    clusters_histogram =\
+        [(x["prediction"], x["count"]) for x in cluster_population]
+
+    # Sort in-place from highest to lowest populated cluster.
+    clusters_histogram.sort(key=lambda x: x[0], reverse=False)
+
+    # Save the cluster ids and their respective scores separately.
+    clusters = [cluster_id for cluster_id, _ in clusters_histogram]
+    counts = [donor_count for _, donor_count in clusters_histogram]
+
+    # Compute the proportion of user in each cluster.
+    total_donors_in_clusters = sum(counts)
+    clust_sample = [float(t) / total_donors_in_clusters for t in counts]
+    sampling_proportions = dict(zip(clusters, clust_sample))
+
+    # Sample the users in each cluster according to the proportions
+    # and pass along the random seed if needed for tests.
+    sampling_kwargs = {"seed": random_seed} if random_seed else {}
+    donor_df = clusters_df.sampleBy("prediction",
+                                    fractions=sampling_proportions,
+                                    **sampling_kwargs)
+    # Get the specific number of donors for each cluster and drop the
+    # predicted cluster number information.
+    current_sample_size = donor_df.count()
```

Comment:
Can we add logging here to check if the actual number of returned donors is close to the requested number of donors?

Reply:
@mlopatka yes, I will add the logging line here. I have a question though: should we make the job fail if we don't get enough donors?

Reply:
As we just discusssed, we need a hard-coded failure condition for the extreme case.
The threshold should be set to <100.

Likewise, if a number of donoros less than 100 is requested at any funciton call (as input param) the job shoudl fail as well.